kedro_mlflow/io/mlflow_metrics_dataset.py

```diff
@@ -0,0 +1,95 @@
+import json
+import os
+from pathlib import Path, PurePosixPath
+from typing import Any, Dict, List, Optional, Union
+
+import mlflow
+from cachetools import Cache
+from kedro.io import AbstractVersionedDataSet
+from kedro.io.core import DataSetError, Version
+
+
+class MlflowMetricsDataSet(AbstractVersionedDataSet):
+    """This class represent mlflow metrics.
+
+    It decorates their ``save`` method to log the metrics in mlflow when
+    ``save`` is called.
+    """
+
+    DEFAULT_SAVE_ARGS: Dict[str, Any] = {"indent": 2}
+
+    def __init__(
+        self,
+        prefix: Optional[str] = None,
+        filepath: Optional[Union[str, PurePosixPath]] = None,
+        version: Optional[Version] = None,
+    ):
+        """Initialise MLflowMetricsDataSet.
+
+        Args:
+            prefix (Optional[str]): Prefix for metrics logged in mlflow.
+            filepath (Optional[Union[str, Path]]): Path to file where dataset
+                will be saved. If not set metrics will be log in mlflow and
+                will not be saved on disk.
+            version (Optional[Version]): If specified, should be an instance of
+                ``kedro.io.core.Version``. If its ``load`` attribute is
+                None, the latest version will be loaded. If its ``save``
+                attribute is None, save version will be autogenerated.
+        """
+        self._filepath = PurePosixPath(filepath) if filepath else filepath
+        self.prefix = prefix
+        self._version = version
+
+        # 1 entry for load version, 1 for save version
+        self._version_cache = Cache(maxsize=2)
+
+    def _load(self) -> Dict[str, Union[float, List[float]]]:
+        """Load MlflowMetricDataSet.
+
+        Returns:
+            Dict[str, Union[int, float]]: Dictionary with MLflow metrics.
+        """
+        if self._filepath:
+            with open(self._get_load_path(), "r") as f:
+                return json.load(f)
+        raise FileNotFoundError(
+            "Filepath parameter is not specified, cannot load dataset."
+        )
+
+    def _save(self, data: Dict[str, Union[float, List[float]]]) -> None:
+        """Save given dataset and log it in mlflow as metrics.
+
+        Args:
+            data (data: Dict[str, float]): MLflow metrics data.
+        """
+        for key, value in data.items():
+            if self.prefix:
+                key = f"{self.prefix}.{key}"
+            if isinstance(value, list):
+                for x in value:
+                    mlflow.log_metric(key, x)
+            else:
+                mlflow.log_metric(key, value)
+
+        if self._filepath:
+            filepath = self._get_save_path()
+            os.makedirs(filepath.parent, exist_ok=True)
+            with open(filepath, "w") as f:
+                json.dump(data, f, **self.DEFAULT_SAVE_ARGS)
```

Comment:
Do you see a possibility to let the user decide whether metrics should be stored as JSON or CSV? This is up for a discussion.

Reply:
Implemented. I like that idea :+1: